src/vsr/superblock_free_set_encoded.zig

```diff
@@ -0,0 +1,449 @@
+
+            assert(reference.size % @sizeOf(FreeSet.Word) == 0);
+            assert(set.size == 0);
+            assert(set.size_transferred == 0);
+            assert(set.block_count == 0);
+            assert(set.block_index == 0);
+
+            set.size = reference.size;
+            set.callback = .{ .open = callback };
+
+            set.block_count = stdx.div_ceil(set.size, chunk_size_max);
+            // Start from the last block, as the linked list arranges data in the reverse order.
+            set.block_index = set.block_count;
+
+            if (set.size == 0) {
+                assert(reference.last_block_address == 0);
+                set.grid.?.on_next_tick(open_next_tick, &set.next_tick);
+            } else {
+                assert(reference.last_block_address != 0);
+                set.open_read_next(reference.last_block_address, reference.last_block_checksum);
+            }
+        }
+
+        fn open_next_tick(next_tick: *Grid.NextTick) void {
+            const set = @fieldParentPtr(Self, "next_tick", next_tick);
+            assert(set.callback == .open);
+            assert(set.block_count == 0);
+            assert(set.size == 0);
+            set.open_done();
+        }
+
+        fn open_read_next(set: *Self, address: u64, checksum: u128) void {
+            assert(set.callback == .open);
+            assert(set.size > 0);
+            assert(address != 0);
+            assert((set.size_transferred == 0) == (set.block_index == set.block_count));
+
+            assert(set.block_index <= set.block_count);
+            assert(set.block_index > 0);
+            set.block_index -= 1;
+
+            set.block_addresses[set.block_index] = address;
+            set.block_checksums[set.block_index] = checksum;
+
+            set.grid.?.read_block(
+                .{ .from_local_or_global_storage = open_read_next_callback },
+                &set.read,
+                address,
+                checksum,
+                .{ .cache_read = true, .cache_write = false },
+            );
+        }
+
+        fn open_read_next_callback(read: *Grid.Read, block: Grid.BlockPtrConst) void {
+            const set = @fieldParentPtr(Self, "read", read);
+            assert(set.callback == .open);
+            assert(set.size > 0);
+            assert(set.block_index < set.block_count);
+
+            const encoded_words = schema.FreeSetNode.encoded_words(block);
+            const chunk = Chunk.for_block(.{
+                .block_index = set.block_index,
+                .block_count = set.block_count,
+                .free_set_size = set.size,
+            });
+
+            stdx.copy_disjoint(
+                .exact,
+                u8,
+                set.buffer[chunk.start..chunk.end],
+                encoded_words,
+            );
+            set.size_transferred += chunk.size;
+
+            if (schema.FreeSetNode.previous(block)) |previous| {
+                assert(set.block_index > 0);
```

Comment:
Assert that `previous.{address,checksum}` is not equal to the `last_block_address`/`last_block_checksum`.